File: week4/Assignment/Question2.py
# import sys
# sys.stdin = open('input.txt', 'r')
# sys.stdout = open('output.txt', 'w')


class DSU:
    def __init__(self, n):
        self.parent = [i for i in range(n+1)]
        # self.array=[i for i in range(n+1)]
        self.rank = [0]*n
        self.number_of_sets = n

    def findset(self, u):
        if self.parent[u] == u:
            return u
        else:
            self.parent[u] = self.findset(self.parent[u])
            return self.parent[u]

    def unions(self, u, v):
        u, v = self.findset(u), self.findset(v)
        if u != v:
            self.parent[u] = v
            self.number_of_sets -= 1

    def total_sets(self):
        return self.number_of_sets


t = int(input())
while(t):
    t -= 1
    n, m = input().strip().split()
    n, m = int(n), int(m)
    answer = DSU(n)
    for i in range(m):
        u, v = input().strip().split()
        u, v = int(u), int(v)
        answer.unions(u, v)
    print(answer.total_sets())

File: week4/Assignment/Question2.cpp
#include <bits/stdc++.h>
using namespace std;
#define vi vector<int>
typedef long long ll;

class DSU
{
private:
    vi parent;
    int number_of_sets;

public:
    DSU(int n)
    {
        for (int i = 0; i <= n; i++)
            parent.push_back(i);
        number_of_sets = n;
    }

    int findset(int u)
    {

        if (parent[u] == u)
            return u;
        else
        {
            parent[u] = findset(parent[u]);
            return parent[u];
        }
    }

    void unions(int u, int v)
    {
        u = findset(u);
        v = findset(v);
        if (u != v)
        {
            parent[u] = v;
            number_of_sets--;
        }
    }
    int total_sets() { return number_of_sets; }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    int t;
    cin >> t;
    while (t--)
    {

        int n, m;
        cin >> n >> m;
        DSU answer(n);
        for (int i = 0; i < m; i++)
        {

            int u, v;
            cin >> u >> v;
            answer.unions(u, v);
        }
        cout << answer.total_sets() << "\n";
    }
    return 0;
}